core/app/models/spree/address.rb

```diff
@@ -97,49 +99,70 @@ def require_zipcode?
     end
 
     private
-      def state_validate
-        # Skip state validation without country (also required)
-        # or when disabled by preference
-        return if country.blank? || !Spree::Config[:address_requires_state]
-        return unless country.states_required
-
-        # ensure associated state belongs to country
-        if state.present?
-          if state.country == country
-            self.state_name = nil #not required as we have a valid state and country combo
+
+    def clear_state_entities
+      clear_state
+      clear_state_name
+    end
+
+    def clear_state
+      self.state = nil
+    end
+
+    def clear_state_name
+      self.state_name = nil
+    end
+
+    def clear_invalid_state_entities
+      if state.present? && (state.country != country)
+        clear_state
+      elsif state_name.present? && !country.states_required? && country.states.empty?
+        clear_state_name
+      end
+    end
+
+    def state_validate
```

Comment:
@krzysiek1507  Yeah! It's very complicated. Will try to fix it in a separate PR a lot of features are associated with it.